File: bisect.c
#include<stdio.h>
#include<math.h>

#define NMAX 100
#define TOL 1e-8
#define X1 -2
#define X2 2

double absolute(double x)	{
	if (x < 0)
		return -1 * x;
	else
		return x;
}

double function(double x)	{
	return exp(-1 * x) + cos(x);
}

double average(double a, double b)	{
	return (a + b) / 2;
}

int main()	{
	double x1, x2, x;
	double f1, f2, fx;

	double absolute(double);
	double function(double);
	double average(double, double);

	printf("This program will find root of given function using Bisection Method\n");
	printf("Please enter lower bound and upper bound.\n");
	scanf("%f %f", &x1, &x2);

	f1 = function(x1);
	printf("%g\n", x1);
	printf("%g\n", f1);
	f2 = function(x2);
	printf("%g\n", x2);
	printf("%g\n", f2);

/****
	if (f1 * f2 < 0)	{
		for (int i = 0; i < NMAX; ++i)	{
			x = average(x1, x2);
			fx = function(x);
			if (fabs (fx) < TOL)
				break;
			if (function(x) * function(x1) < 0)
				x2 = x;
			else
				x1 = x;
		}
		printf("%g is root\n", x);
	}
	else if (function(x1) == 0)
		printf("%g is root\n", x1);
	else if (function(x2) == 0)
		printf("%g is root\n", x2);
	else
		printf("Select different interval");
****/

return 0;
}

File: assignment0/bisect.f90
! This Program will find root of equation using bisection method
! Langauage : Fortran
! Author : Keshav Sishodia
! Rollno : PH18D204
! You can use and modify this program as much you want.

real function f(x)
        implicit none
        real, intent(in) :: x
        f = exp(-1 * x) + cos(x)
!        f = 10*x*exp(-1 * x * x) - 1
end function f

program bisect
        implicit none
        integer :: iter, nmax
        real :: x0, x1, x, tol
        real :: f0, f1, fx

        real :: f

        print *, 'This Program will find root of equation using bisection method.'
        print *, 'Please enter lower bound, upper bound and tolerence.'
        read (*,*) x0, x1, tol
        print *, "Enter maximum no of iteration you want to try."
        read (*,*) nmax
        tol = abs(tol)
        
        f0 = f(x0); f1 = f(x1)

        if (f0 * f1 < 0)        then
                do iter=0,nmax,1
                        if (iter == nmax)       then
                                print *, "No if iteration has exceded than your defined maximum."
                        endif
                        x = (x0 + x1) / 2
                        fx = f(x)
                        if (abs(fx) <= tol)    then
                                print *, "After iteration", iter
                                print *, "Root of equation is :", x
                                exit
                        else if (f0 * fx < 0)   then
                                x1 = x
                        else
                                x0 = x
                                f0 = fx
                        endif
                enddo
        else if (abs(f0) < tol) then
                print *, "Root of equation is :", x0
        else if (abs(f1) < tol) then
                print *, "Root of equation is :", x1
        else
                print *, "Root does not lie in this interval"
        endif

end program bisect
